src/natpmp.cpp

```diff
@@ -344,15 +478,73 @@ void natpmp::send_map_request(port_mapping_t const i)
 	m_currently_mapping = i;
 	mapping_t& m = m_mappings[i];
 	TORRENT_ASSERT(m.act != portmap_action::none);
-	char buf[12];
+	char buf[60];
 	char* out = buf;
-	write_uint8(0, out); // NAT-PMP version
-	write_uint8(m.protocol == portmap_protocol::udp ? 1 : 2, out); // map "protocol"
-	write_uint16(0, out); // reserved
-	write_uint16(m.local_port, out); // private port
-	write_uint16(m.external_port, out); // requested public port
 	int ttl = m.act == portmap_action::add ? 3600 : 0;
-	write_uint32(ttl, out); // port mapping lifetime
+	if (m_version == version_natpmp)
+	{
+		write_uint8(m_version, out);
+		write_uint8(m.protocol == portmap_protocol::udp ? 1 : 2, out); // map "protocol"
+		write_uint16(0, out); // reserved
+		write_uint16(m.local_port, out); // private port
+		write_uint16(m.external_port, out); // requested public port
+		write_uint32(ttl, out); // port mapping lifetime
+	}
+	else if (m_version == version_pcp)
+	{
+		// PCP requires the use of IPv6 addresses even for IPv4 messages
+		// reference asio's address_v6 class directly so that we can use it
+		// even if TORRENT_USE_IPV6 is false
+		using boost::asio::ip::address_v6;
+		write_uint8(m_version, out);
+		write_uint8(opcode_map, out);
+		write_uint16(0, out); // reserved
+		write_uint32(ttl, out);
+		address const local_addr = m_socket.local_endpoint().address();
+		auto const local_bytes = local_addr.is_v4()
+			? address_v6::v4_mapped(local_addr.to_v4()).to_bytes()
+			: local_addr.to_v6().to_bytes();
+		out = std::copy(local_bytes.begin(), local_bytes.end(), out);
+		out = std::copy(m.nonce.begin(), m.nonce.end(), out);
+		// translate portmap_protocol to an IANA protocol number
+		int protocol = 0;
+		switch (m.protocol)
+		{
+			case portmap_protocol::tcp:protocol = 6; break;
+			case portmap_protocol::udp:protocol = 17; break;
+			case portmap_protocol::none:protocol = 0; break;
+		}
```

Comment:
this could be a bit shorter, and `protocol` could be const if it's initialized with the ternary operator instead.

```
int const protocol =
   (m.protocol == portmap_protocol::tcp) ? 6
   : (m.protocol == portmap_protocol::udp) ? 17
   : 0;
```